packages/hash/frontend/src/blocks/userBlocks.tsx

```diff
@@ -0,0 +1,119 @@
+import { BlockMetadata } from "blockprotocol";
+import produce from "immer";
+import {
+  createContext,
+  Dispatch,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useMemo,
+} from "react";
+import { useLocalstorageState } from "rooks";
+
+export type UserBlock = BlockMetadata & {
+  componentId?: string;
+  packagePath?: string;
+};
+
+export const getComponentId = (meta: UserBlock) => {
+  if (meta.componentId) {
+    return meta.componentId;
+  }
+
+  if (meta.packagePath) {
+    return `https://blockprotocol.org/blocks/${meta.packagePath}`;
+  }
+
+  throw new Error("Added a block without packagePath or componentId");
+};
+
+type UserBlocks = UserBlock[];
+
+interface UserBlocksContextState {
+  value: UserBlocks;
+  setValue: Dispatch<SetStateAction<UserBlocks>>;
+}
+
+/** @private enforces use of custom provider */
+const UserBlocksContext = createContext<UserBlocksContextState | null>(null);
+
+const mergeBlocksData = (
+  oldBlocksData: UserBlocks,
+  newBlocksData: UserBlocks,
+): UserBlocks => {
+  return produce(oldBlocksData, (draftUserBlocks) => {
+    for (const latestUserBlock of newBlocksData) {
+      const matchingUserBlockIndex = draftUserBlocks.findIndex(
+        (userBlock) =>
+          userBlock.name === latestUserBlock.name &&
+          userBlock.version !== latestUserBlock.version,
+      );
+
+      // Using `any` to fix a `Type instantiation is excessively deep` error.
+      // @todo find a potential fix.
+      if (matchingUserBlockIndex > -1) {
+        draftUserBlocks[matchingUserBlockIndex] = latestUserBlock as any;
+      } else {
+        draftUserBlocks.push(latestUserBlock as any);
+      }
+    }
+  });
+};
+
+export const UserBlocksProvider: React.FC<{ value: UserBlocks }> = ({
+  value: initialUserBlocks,
+  children,
+}) => {
+  const [value, setValue] = useLocalstorageState<UserBlocks>(
+    "hash-workspace-user-blocks",
+    initialUserBlocks,
+  );
+
+  useEffect(() => {
+    const setInitialBlocks = async () => {
+      if (process.env.NEXT_PUBLIC_BLOCK_PROTOCOL_API_KEY) {
+        const blocksMetadataHeaders = new Headers();
+        blocksMetadataHeaders.set(
+          "x-api-key",
+          process.env.NEXT_PUBLIC_BLOCK_PROTOCOL_API_KEY,
+        );
+
+        const userBlocks = (
+          await fetch("https://blockprotocol.org/api/blocks", {
+            headers: blocksMetadataHeaders,
+          }).then((response) => response.json())
+        ).results as UserBlocks;
+
+        setValue((prevValue) => {
+          return mergeBlocksData(prevValue, userBlocks);
+        });
+      }
+    };
+
+    void setInitialBlocks();
```

Comment:
@MaggieAppleton Please can you come up with a state for this? Probably some in-menu warning that "Not all blocks are available right now. Please choose from the list below." with some subset, or similar

Reply:
Is there a possibility no blocks are cached? Eg. this is the first time the user has used the app and the API isn't working – could the list could be empty?

Reply:
For sure yeah

Reply:
@akash-joshi Here's mockups for the loading state error
![image](https://user-images.githubusercontent.com/5599295/159691507-1786fcc7-a5a0-4523-a8c0-04b572bd2f6f.png)

figma: https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-Editor?node-id=1979%3A109853

Reply:
Here's what it looks like, based on Maggie's designs.
Caveats:
1. I'm on Windows so the scrollbar is quite ugly. It looked the same even before these changes.
2. The Table block icon looks broken because it's pointing to my locally served block (which I hadn't brought up, hence the broken icon).

![chrome_57yuDDawNI](https://user-images.githubusercontent.com/22196279/159879341-d2c0ea6d-1af6-4737-af0e-8ee9e178d835.gif)